rust/blockstore/src/arrow/block/types.rs

```diff
@@ -118,25 +119,35 @@ impl Block {
         delta
     }
 
-    /// Binary search the blockfile to find the partition point of the specified prefix and key.
-    /// The implementation is based on [`std::slice::partition_point`].
+    /// Binary searches this slice with a comparator function.
     ///
-    /// `(prefix, key)` serves as the search key, and it is sorted in ascending order.
-    /// The partition predicate is defined by: `|x| x < (prefix, key)`.
-    /// The partition point is the first index where the partition precidate evaluates to `false`
-    /// The code is a result of inlining this predicate in [`std::slice::partition_point`].
-    /// If the key is unspecified (i.e. `None`), we find the first index of the prefix.
+    /// The comparator function should return an order code that indicates
+    /// whether its argument is `Less`, `Equal` or `Greater` the desired
+    /// target.
+    /// If the slice is not sorted or if the comparator function does not
+    /// implement an order consistent with the sort order of the underlying
+    /// slice, the returned result is unspecified and meaningless.
     ///
-    /// [`std::slice::partition_point`]: std::slice::partition_point
+    /// If the value is found then [`Result::Ok`] is returned, containing the
+    /// index of the matching element. If there are multiple matches, then any
+    /// one of the matches could be returned. The index is chosen
+    /// deterministically, but is subject to change in future versions of Rust.
+    /// If the value is not found then [`Result::Err`] is returned, containing
+    /// the index where a matching element could be inserted while maintaining
+    /// sorted order.
+    ///
+    /// Based on std::slice::binary_search_by with minimal modifications (https://doc.rust-lang.org/src/core/slice/mod.rs.html#2770).
     #[inline]
-    fn binary_search_index<'me, K: ArrowReadableKey<'me>>(
+    fn binary_search_by<'me, K: ArrowReadableKey<'me>, F>(
```

Comment:
Nice! Curious, why does f have to be `FnMut` as opposed to `Fn`. Can't it work on immutable references?

Reply:
I believe that's based on the standard library implementation. Probably that leaves more flexibility for what can be passed in as a comparator function.